### PR TITLE
Add _type for all parsers

### LINKED ISSUE #3217: Bug in saving agent since version v0.0.142
I've been investigating an error when running agent based example with the Comet callback when trying to save the agent to disk.

I have been able to narrow the bug down to the following reproduction script:

```python
import os
from datetime import datetime

import langchain
from langchain.agents import initialize_agent, load_tools
from langchain.callbacks import CometCallbackHandler, StdOutCallbackHandler
from langchain.callbacks.base import CallbackManager
from langchain.llms import OpenAI

llm = OpenAI(temperature=0.9, verbose=True)

tools = load_tools(["serpapi", "llm-math"], llm=llm)
agent = initialize_agent(
    tools,
    llm,
    agent="zero-shot-react-description",
    verbose=True,
)

agent.save_agent("/tmp/langchain.json")
```

Which fails with the following Exception:
```python
Traceback (most recent call last):
  File "/home/lothiraldan/project/cometml/langchain/docs/ecosystem/test_comet_ml_3.py", line 39, in <module>
    agent.save_agent("/tmp/langchain.json")
  File "/home/lothiraldan/project/cometml/langchain/langchain/agents/agent.py", line 599, in save_agent
    return self.agent.save(file_path)
  File "/home/lothiraldan/project/cometml/langchain/langchain/agents/agent.py", line 145, in save
    agent_dict = self.dict()
  File "/home/lothiraldan/project/cometml/langchain/langchain/agents/agent.py", line 119, in dict
    _dict = super().dict()
  File "pydantic/main.py", line 435, in pydantic.main.BaseModel.dict
  File "pydantic/main.py", line 833, in _iter
  File "pydantic/main.py", line 708, in pydantic.main.BaseModel._get_value
  File "/home/lothiraldan/project/cometml/langchain/langchain/schema.py", line 381, in dict
    output_parser_dict["_type"] = self._type
  File "/home/lothiraldan/project/cometml/langchain/langchain/schema.py", line 376, in _type
    raise NotImplementedError
NotImplementedError
```

Using that reproduction script, I was able to run git bisect that identified the following commit as the probable cause: https://github.com/hwchase17/langchain/commit/e12e00df12c6830cd267df18e96fda1ef8df6c7a

I am not sure of the scope of that issue and if it's mean that no agent can be exported to JSON or YAML since then.

Let me know if I can help more on debugging that issue.

### PR DESCRIPTION
**Pull request by sunishsheth2009**
Used for serialization. Also add test that recurses through
our subclasses to check they have them implemented

Would fix https://github.com/hwchase17/langchain/issues/3217
Blocking: https://github.com/mlflow/mlflow/pull/8297

### CODE REVIEW

**Comment by sunishsheth2009** [comment]
Replicating the PR: https://github.com/hwchase17/langchain/pull/3332
Thank you for your help: @vowelparrot , happy to close this PR in favor of the original PR :)

File: langchain/output_parsers/fix.py [hunk 1]
```diff
@@ -45,4 +45,4 @@ def get_format_instructions(self) -> str:
 
     @property
     def _type(self) -> str:
-        return self.parser._type
+        return f"output_fixing_{self.parser._type}"
```
Comment (dev2049) [inline langchain/output_parsers/fix.py:48]: let's keep this as a constant (just "output_fixing" is fine). the input parser will be serialized itself, so no need to encode information about it here. same thing for RetryOutputParser

File: tests/unit_tests/output_parsers/test_base_output_parser.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from unittest.mock import MagicMock
+
+import pytest
+
+from langchain.schema import BaseOutputParser
+
+_PARSERS_TO_SKIP = {"FakeOutputParser", "BaseOutputParser"}
+
+
+def non_abstract_subclasses(cls: Type[ABC]) -> List[Type]:
```
Comment (dev2049) [inline tests/unit_tests/output_parsers/test_base_output_parser.py:13]: small thing, but given that this fn is fully generic aside from PARSERS_TO_SKIP, maybe we could instead add a `to_skip: Optional[Set]=None` parameter and pass _PARSERS_TO_SKIP in instead of hardcoding in this fn

**Review by dev2049 — COMMENTED** [review]
2 small comments, otherwise looks good!

File: tests/unit_tests/output_parsers/test_base_output_parser.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        subclasses.extend(non_abstract_subclasses(subclass, to_skip))
+    return subclasses
+
+
+@pytest.mark.parametrize("cls", non_abstract_subclasses(BaseOutputParser, _PARSERS_TO_SKIP))
+def test_all_subclasses_implement_type(cls: Type[BaseOutputParser]) -> None:
+    try:
+        # Most parsers just return a string. MagicMock lets
+        # the parsers that wrap another parsers slide by
+        cls._type.fget(MagicMock())  # type: ignore
```
Comment (dev2049) [inline tests/unit_tests/output_parsers/test_base_output_parser.py:29]: not sure i understand, why is this needed?

File: tests/unit_tests/output_parsers/test_base_output_parser.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    for subclass in cls.__subclasses__():
+        if not getattr(subclass, "__abstractmethods__", None):
+            if subclass.__name__ not in to_skip:
+                subclasses.append(subclass)
+        subclasses.extend(non_abstract_subclasses(subclass, to_skip))
+    return subclasses
+
+
+@pytest.mark.parametrize("cls", non_abstract_subclasses(BaseOutputParser, _PARSERS_TO_SKIP))
+def test_all_subclasses_implement_type(cls: Type[BaseOutputParser]) -> None:
```
Comment (dev2049) [inline tests/unit_tests/output_parsers/test_base_output_parser.py:25]: very small nit: test_subclass_implements_type is slightly more appropriate name given this asserts over a single class at a time

File: tests/unit_tests/output_parsers/test_base_output_parser.py (hunk 3 above)
Comment (sunishsheth2009) [reply tests/unit_tests/output_parsers/test_base_output_parser.py:29]: I agree, its not needed here. There are no parsers that wrap another parser. Removed that code. :) 
Thank you
